src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -874,46 +872,106 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           )
       }
 
-      def instantiateToMethodType(mt: MethodType): Tree = {
-        val meth = tree match {
-          // a partial named application is a block (see comment in EtaExpansion)
-          //
-          // TODO: document why we need to peel off one layer to begin with, and, if we do, why we don't need to recurse??
-          // (and why we don't need to look at the stats to make sure it's the block created by eta-expansion)
-          // I don't see how we call etaExpand on a constructor..
-          //
-          // I guess we don't need to recurse because the outer block and a nested block will refer to the same method
-          // after eta-expansion (if that's what generated these blocks), so we can just look at the outer one.
-          //
-          // How about user-written blocks? Can they ever have a MethodType?
-          case Block(_, tree1) => tree1.symbol
-          case _               => tree.symbol
+      def adaptMethodTypeToExpr(mt: MethodType): Tree = {
+        val meth =
+          tree match {
+            // a partial named application is a block (see comment in EtaExpansion)
+            // How about user-written blocks? Can they ever have a MethodType?
+            case Block(_, tree1) => tree1.symbol
+            case _               => tree.symbol
+          }
+
+
+        val arity = mt.params.length
+
+        val sourceLevel2_14 = currentRun.isScala214
+
+        def warnTree = original orElse tree
+
+        def warnEtaSam() = {
+          val sam = samOf(pt)
+          if (sam.exists && !sam.owner.hasAnnotation(definitions.FunctionalInterfaceClass))
+            reporter.warning(tree.pos, s"Eta-expansion performed to meet expected type $pt, which is SAM-equivalent to ${samToFunctionType(pt)},\n" +
+                                       s"even though ${sam.owner} is not annotated with `@FunctionalInterface`; add annotation to suppress warning.")
```

Comment:
Can this code path occur within silent-mode typechecking? `reporter.warning` will issue this warning even if this line of inquiry is discarded.

Reply:
The advice to annotate with `@FunctionalInterface` is only valid if:

  - The type is under the user's control, not third party
  - It is not a class in a .java file (as javac only allows it on interfaces)

So the warning should probably be nuanced.

Another possible recommendation would be to explicitly eta-expand in source.

Reply:
Fixed wording and only trigger on interface, if Java-defined. Also (pro/de)-moted to lint warning.